aten/src/ATen/native/mkldnn/xpu/detail/Attention.cpp

```diff
@@ -0,0 +1,294 @@
+#include <ATen/native/mkldnn/xpu/detail/Attr.h>
+#include <ATen/native/mkldnn/xpu/detail/Graph.h>
+#include <ATen/native/mkldnn/xpu/detail/Utils.h>
+#include <ATen/native/mkldnn/xpu/detail/oneDNN.h>
+
+#include <omp.h>
+#include <oneapi/dnnl/dnnl.hpp>
+
+using namespace at::native::onednn::graph;
+namespace {
+struct SDPALogicalParams {
+  enum class TensorID {
+    query,
+    key,
+    scale,
+    attn_mask,
+    value,
+    output,
+    end,
+  };
+
+  logical_tensor query;
+  logical_tensor key;
+  logical_tensor scale;
+  std::optional<logical_tensor> attn_mask;
+  logical_tensor value;
+  logical_tensor output;
+
+  SDPALogicalParams(
+      const at::Tensor& query_,
+      const at::Tensor& key_,
+      const at::Tensor& value_,
+      const std::optional<at::Tensor>& attn_mask_,
+      const at::Tensor& output_,
+      data_type dtype) {
+    dims scale_shape = {1};
+    std::vector<logical_tensor> inputLogicalTensors;
+    query = {
+        static_cast<size_t>(TensorID::query),
+        dtype,
+        query_.sizes().vec(),
+        query_.strides().vec()};
+    key = {
+        static_cast<size_t>(TensorID::key),
+        dtype,
+        key_.sizes().vec(),
+        key_.strides().vec()};
+    scale = {
+        static_cast<size_t>(TensorID::scale),
+        dtype,
+        scale_shape,
+        logical_tensor::layout_type::strided,
+        logical_tensor::property_type::constant};
+    if (attn_mask_.has_value()) {
+      attn_mask = {
+          static_cast<size_t>(TensorID::attn_mask),
+          dtype,
+          attn_mask_->sizes().vec(),
+          attn_mask_->strides().vec()};
+    }
+    value = {
+        static_cast<size_t>(TensorID::value),
+        dtype,
+        value_.sizes().vec(),
+        value_.strides().vec()};
+    output = {
+        static_cast<size_t>(TensorID::output),
+        dtype,
+        output_.sizes().vec(),
+        output_.strides().vec()};
+  }
+  std::vector<logical_tensor> get_input() const {
+    if (attn_mask.has_value()) {
+      return {query, key, scale, attn_mask.value(), value};
+    } else {
+      return {query, key, scale, value};
+    }
+  }
+  std::vector<logical_tensor> get_output() const {
+    return {output};
+  }
+};
+
+partition create_sdpa_graph_partition(
+    int batch_size,
+    int seq_len_q,
+    int seq_len_k,
+    int num_head,
+    int head_dim,
+    bool is_causal,
+    data_type dtype,
+    const SDPALogicalParams& params) {
+  // graph building and partitioning
+  // currently, we assume that Q and K have same sequence length
+
+  dims qk_output_shape = {batch_size, num_head, seq_len_q, seq_len_k};
+  dims scale_shape = {1};
+  size_t lt_id = static_cast<size_t>(SDPALogicalParams::TensorID::end);
+  size_t op_id = 0;
+
+  logical_tensor matmul_qk_out{lt_id++, dtype};
+  op matmul_qk{
```

Comment:
usually we should skip using `using namespace at::native::onednn::graph;` therefore the rest of the code would be difficult to read. here it is better to put `onednn::graph::op` or you may simplify it as you like, but do not just put `op` here.